TPM attestation fixes

tomoveu opened the pull request:
The time print of the signed timestamp had to be unmarshalled(parsed) after the TPM gave us the response.

It is a matter of using the correct structure type and indeed the cast had to be corrected to the TPMS_ATTEST structure.

However, this exposed another issue with marshalling/unmarshalling(append/parse) that I will rise as an issue.

In nutshell, the safe-check with TPM_GENERATED_VALUE shows that our parsed buffer has the data in Big Endian format, i.e the TPM buffer has not been completely parsed(unmarshalled). More information in #104 

Signed-off-by: Dimitar Tomov <dimi@designfirst.ee>

dgarske commented:
@tomoveu : I added attestation parsing to this PR. Please review and let me know your thoughts.

embhorn commented on src/tpm2_packet.c:
```diff
@@ -725,6 +725,95 @@ void TPM2_Packet_ParseSignature(TPM2_Packet* packet, TPMT_SIGNATURE* sig)
+
+    TPM2_Packet_ParseU32(packet, &out->magic);
+    if (out->magic != TPM_GENERATED_VALUE)
+        return;
```
A debug message in case of failure here would be useful.

tomoveu commented:
@dgarske the new function for parsing the TPM attest structures with endianness is looking good. Two small but important changes from me. Two details.

1. A member was missing for the TPMS_ATTEST_TIME_INFO parsing, which leads to false printing. Quick fix, 'time' is the first U64 value, before the TPM clock. 

2. Then, there is an important difference in the purpose of TPM time and TPM clock. Let me elaborate:

**TPMS_TIME_ATTEST_INFO** is the attested data returned by GetTime, this holds TPMS_TIME_INFO and firmwareVersion members.

**TPMS_TIME_INFO** holds the special 'time' member and standard TPMS_CLOCK_INFO

From the TCG spec: 

> The Time  value  reported  in  this  structure  is  reset  whenever  the  TPM  is  reset.

TPMS_CLOCK_INFO

From the TCG spec:

> Clock  is  a  monotonically  increasing  counter  that  advances  whenever  power  is  applied  to  the  TPM.

And I saw the print for TPM time was removed, so I added back the TPM time print and modified the messages so they clearly explain the output data ;-)

Other than these two details, the unmarshalling(parsing) function for Attested structures looks really good and I can confirm that it works perfectly for GetTime/TPMS_TIME_ATTEST_INFO. The other attest structures(cases) we will have to test as we make them available for use.

Good job, David. Thank you for the quick reaction. Now the signed_timestamp example is bullet-proof :100:

tomoveu commented:
> Very nice!
> Does this value for `Restart Count` sound plausible?
> 
> ```
> wolfTPM $ ./examples/timestamp/signed_timestamp 
> TPM2 Demo of generating signed timestamp from the TPM
> 	Restart Count: 3164145585
> 	Clock Safe: 0
> 	Firmware Version: 1
> ```

No, there was a member missed in the parsing of the TPMS_TIME_ATTEST_INFO, which caused for this. I just pushed new changes to adjust for that. Also highlighting the difference between TPM time and TPM clock.

Here are two consecutive prints to illustrate:
```

trm@PCMXPS:~/wolfTPM$ sudo examples/timestamp/signed_timestamp 
TPM2 Demo of generating signed timestamp from the TPM
wolfTPM2_Init: success
TPM2_ReadClock: success
wolfTPM2_CreateEK: Endorsement 0x80000000 (314 bytes)
wolfTPM2_CreateSRK: Storage 0x80000001 (282 bytes)
TPM2_StartAuthSession: sessionHandle 0x3000000
TPM2_policySecret success
wolfTPM2_CreateAndLoadAIK: AIK 0x80000002 (280 bytes)
wolfTPM2_GetTime: success
TPM with signature attests (type 0x8019):
	TPM uptime since power-up(in ms): 830641
	TPM clock, total time the TPM has been on(in ms): 425288984
	Reset Count: 30
	Restart Count: 0
	Clock Safe: 1
	Firmware Version(vendor specific): 0x1000300010000
trm@PCMXPS:~/wolfTPM$ sudo examples/timestamp/signed_timestamp 
TPM2 Demo of generating signed timestamp from the TPM
wolfTPM2_Init: success
TPM2_ReadClock: success
wolfTPM2_CreateEK: Endorsement 0x80000000 (314 bytes)
wolfTPM2_CreateSRK: Storage 0x80000001 (282 bytes)
TPM2_StartAuthSession: sessionHandle 0x3000000
TPM2_policySecret success
wolfTPM2_CreateAndLoadAIK: AIK 0x80000002 (280 bytes)
wolfTPM2_GetTime: success
TPM with signature attests (type 0x8019):
	TPM uptime since power-up(in ms): 851276
	TPM clock, total time the TPM has been on(in ms): 425309619
	Reset Count: 30
	Restart Count: 0
	Clock Safe: 1
	Firmware Version(vendor specific): 0x1000300010000
trm@PCMXPS:~/wolfTPM$ 

```

dgarske commented:
@tomoveu : Dimi you are welcome!

@embhorn , here is how mine looks now after Dimi's fixes.

```
TPM with signature attests (type 0x8019):
	TPM uptime since last power-up(in ms): 16482
	TPM clock, total time the TPM has been on(in ms): 1296901257
	Reset Count: 47
	Restart Count: 0
	Clock Safe: 1
	Firmware Version(vendor specific): 0x11CB00
```

tomoveu commented:
> @tomoveu : Dimi you are welcome!
> 
> @embhorn , here is how mine looks now after Dimi's fixes.
> 
> ```
> TPM with signature attests (type 0x8019):
> 	TPM uptime since last power-up(in ms): 16482
> 	TPM clock, total time the TPM has been on(in ms): 1296901257
> 	Reset Count: 47
> 	Restart Count: 0
> 	Clock Safe: 1
> 	Firmware Version(vendor specific): 0x11CB00
> ```

looks perfect :+1: 

Make two consecutive executions to see the increment in TPM time (uptime since last power-up) and TPM clock (total time the TPM was ever powered). Very nice demo. Also, shows how the vendor string, reset and restart count stay the same. 

Teamwork FTW :1st_place_medal:
